cryptography/src/transcript.rs

```diff
@@ -0,0 +1,107 @@
+use rand_core::{CryptoRngCore, OsRng};
+
+/// This struct provides a convenient abstraction over hashing data and deriving randomness.
+///
+/// It automatically takes care of details like:
+/// - correctly segmenting packets of data,
+/// - domain separating different uses of tags and randomness,
+/// - making sure that secret state is zeroized as necessary.
+struct Transcript {}
+
+impl Transcript {
+    /// Create a new transcript.
+    ///
+    /// The namespace serves to disamiguate two transcripts, so that even if they record
+    /// the same information, the results will be different:
+    /// ```
+    /// let s1 = Transcript::new(b"n1").record(b"A").summarize();
+    /// let s2 = Transcript::new(b"n2").record(b"A").summarize();
+    /// assert_neq!(s1, s2);
+    /// ```
+    pub fn new(namespace: &[u8]) -> Self {
+        todo!()
+    }
+
+    /// Start a transcript from a summary.
+    ///
+    /// Note that this will not produce the same result as if the transcript
+    /// were never summarized to begin with.
+    /// ```
+    /// let s1 = Transcript::new(b"test").record(b"A").summarize();
+    /// let s2 = Transcript::resume(s1).summarize();
+    /// assert_neq!(s1, s2);
+    /// ```
+    pub fn resume(summary: Summary) -> Self {
+        todo!()
+    }
+
+    /// Record data in this transcript.
+    ///
+    /// Calls to record automatically separate out data:
+    /// ```
+    /// let s1 = Transcript::new(b"test").record(b"A").record(b"B").summarize();
+    /// let s2 = Transcript::new(b"test").record(b"AB").summarize();
+    /// assert_neq!(s1, s2);
+    /// ```
+    ///
+    /// If you want to provide data incrementally, use [Self::record_partial]
+    pub fn record(&mut self, data: &[u8]) -> &mut Self {
+        todo!()
+    }
+
+    /// Like [Self::record], except that subsequent calls are considered part of the same message.
+    ///
+    /// ```
+    /// let s1 = Transcript::new(b"test").record_partial(b"A").record(b"B").summarize();
+    /// let s2 = Transcript::new(b"test").record(b"AB").summarize();
+    /// assert_eq!(s1, s2);
+    /// ```
+    pub fn record_partial(&mut self, data: &[u8]) -> &mut Self {
+        todo!()
+    }
+
+    /// Split this transcript into two independent transcripts.
+    ///
+    /// These transcripts are both linked to the same history, but will produce different
+    /// randomness, and can be updated without affecting the other one.
+    /// This is often useful as a pre-requisite to methods like [Self::summarize] or
+    /// [Self::noise], which consume the transcript. It can be useful to extract
+    /// randomness, or a summary, while still continuing to feed data, and this method
+    /// allows you to do that.
+    pub fn split(self) -> (Self, Self) {
+        todo!()
+    }
+
+    /// Turn this transcript into a source of random bytes.
+    pub fn noise(self) -> impl CryptoRngCore {
+        // TODO: This is just to compile
+        OsRng
+    }
```

Comment:
I sort of think `summarize` should produce this?

If `Transcript` is a trait with associated type `Summary`, you can also require `type Summary: Eq + PartialEq + CryptoRngCore`

Reply:
I think the semantic difference is that `summarize` returns a fixed, static object, essentially a commitment over the transcript so far, whereas `noise` returns a mutable object, which you can pull from as a source of randomness, and mutate. In particular, whether or not two such sources are equal is more complicated. Also, you may not necessarily be able to serialize it, whereas my intention is that you can serialize `Summary`, allowing transcripts to be saved to disk and resumed (among other use cases).